polaris-core/src/main/java/org/apache/polaris/core/persistence/BasePersistence.java

```diff
@@ -0,0 +1,435 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.core.persistence;
+
+import jakarta.annotation.Nonnull;
+import jakarta.annotation.Nullable;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import org.apache.polaris.core.PolarisCallContext;
+import org.apache.polaris.core.entity.PolarisBaseEntity;
+import org.apache.polaris.core.entity.PolarisChangeTrackingVersions;
+import org.apache.polaris.core.entity.PolarisEntityActiveRecord;
+import org.apache.polaris.core.entity.PolarisEntityCore;
+import org.apache.polaris.core.entity.PolarisEntityId;
+import org.apache.polaris.core.entity.PolarisEntityType;
+import org.apache.polaris.core.entity.PolarisGrantRecord;
+import org.apache.polaris.core.entity.PolarisPrincipalSecrets;
+import org.apache.polaris.core.storage.PolarisStorageConfigurationInfo;
+import org.apache.polaris.core.storage.PolarisStorageIntegration;
+
+/**
+ * Interface to the Polaris metadata store, allows to persist and retrieve all Polaris metadata like
+ * metadata for Polaris entities and metadata about grants between these entities which is the
+ * foundation of our role base access control model.
+ *
+ * <p>Note that APIs to the actual persistence store are very basic, often point read or write to
+ * the underlying data store. The goal is to make it really easy to back this using databases like
+ * Postgres or simpler KV store.
+ */
+public interface BasePersistence {
+  /**
+   * @param callCtx call context
+   * @return new unique entity identifier
+   */
+  long generateNewId(@Nonnull PolarisCallContext callCtx);
+
+  /**
+   * Write this entity to the meta store.
+   *
+   * @param callCtx call context
+   * @param entity entity to persist
+   * @param nameOrParentChanged if true, also write it to by-name lookups if applicable
+   * @param originalEntity original state of the entity to use for compare-and-swap purposes, or
```

Comment:
I believe we need to have explicit statements about concurrent update expectations (e.g. exception on conflict) and durability expectation (i.e. successful return means that the change was persisted).

Reply:
Yeah I agree. I should note - most of these comments right now are still just directly copy/pasted from the original `PolarisMetaStoreSession`, so I'm just chipping away at clarifying the interfaces :)

I think the way to handle the write conflict still needs a little tuning which I was going to do in a followup since this PR is able to define the *direction* of the interfaces without actually introducing any behavior changes yet.

I'll add a TODO here to either standardize exceptions for the error cases or push down the return status enum from PolarisMetaStoreManager